### PR TITLE
Upgrade to Ahoy 2.0.x

### PR DESCRIPTION
**Pull request by joshpuetz**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Library Upgrade

## Description

Upgrades the [ahoy_email](https://github.com/ankane/ahoy_email) gem to version 2.0.2. There were a few breaking changes in the upgrade that required this to be a manual upgrade.

The biggest consequence of this upgrade is that [open tracking has been removed](https://github.com/ankane/ahoy_email#upgrading). As a result: `open.gif` is no longer included in emails send by Ahoy, and we can no longer track when users open an email. I'm unclear if there are downstream marketing impact (maybe @pkfrank could chime in)

Another consequence is we can no longer use open frequency as part of a calculation of when to send digest emails. I've replaced the two existing `SiteConfig` values controller min and max email digest send frequency with a single value, updated the admin UI, and changed the digest email sending code to always use this static value.

Change and removal of this gem (as mentioned in https://github.com/forem/forem/pull/13085#issuecomment-805271489) will be tackled in a separate RFC.

## Related Tickets & Documents

https://github.com/forem/forem/pull/13037

## Added tests?

- [ ] Yes
- [x] No, and this is why: existing tests fixed
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/models/email_message.rb [hunk 1]
```diff
@@ -6,8 +6,8 @@ class EmailMessage < Ahoy::Message
   # reasons to define behavior here, similar to how we use the Tag model.
   def body_html_content
     doctype_index = content.index("<!DOCTYPE")
-    closing_html_index = content.index("</html>") + 6
-    content[doctype_index..closing_html_index]
+    closing_body_index = content.index("</body>") + 6
+    content[doctype_index..closing_body_index]
```
Comment (joshpuetz) [inline app/models/email_message.rb:10]: This code had to be changed because Ahoy is no longer doing HTML processing with Nokogiri: it's treating the email as a fragment: https://github.com/ankane/ahoy_email/compare/v1.1.1...v2.0.2#diff-68d3d25e705f575b9bd84cc277469ed00524fa4e38e69d5291310887eea06bcbR57

File: spec/mailers/notify_mailer_spec.rb [hunk 2]
```diff
@@ -24,10 +24,6 @@
       expect(email.to).to eq([comment.user.email])
     end
 
-    it "includes the tracking pixel" do
-      expect(email.html_part.body).to include("open.gif")
-    end
-
```
Comment (joshpuetz) [inline spec/mailers/notify_mailer_spec.rb:30]: Open tracking is no longer supported

**Review by juliannatetreault — APPROVED** [review]
Looks good to me! Thanks for the inline comments. 😄

**Review by jacobherrington — APPROVED** [review]

File: spec/mailers/digest_mailer_spec.rb [hunk 3]
```diff
@@ -21,7 +21,6 @@
 
     it "includes the tracking pixel" do
```
Comment (rhymes) [inline spec/mailers/digest_mailer_spec.rb:22]: I think we can remove this whole test, it has no expectations now

**Review by rhymes — COMMENTED** [review]
Nice work @joshpuetz!

Should we also remove the column `opened_at` from the `ahoy_messages` table ?

File: spec/mailers/digest_mailer_spec.rb (hunk 3 above)
Comment (joshpuetz) [reply spec/mailers/digest_mailer_spec.rb:22]: Wow, how'd I miss that?! 😆

**Comment by joshpuetz** [comment]
> Nice work @joshpuetz!
> 
> Should we also remove the column `opened_at` from the `ahoy_messages` table ?

Only about 1/3 of the existing records on this table have this field populated, so it's in rare usage even now. I'll remove it!

**Comment by djuber** [comment]
still a few references to the dropped `opened_at` column in the failing tests (hooray for coverage)

File: app/services/email_digest_article_collector.rb [hunk 4]
```diff
@@ -49,20 +49,7 @@ def days_until_next_email
...
-
-  def open_rate
-    email_count = last_user_emails.count
-
-    # Will stick with 50% open rate if @user has no/not-enough email digest history
-    return 0.5 if email_count < 10
-
-    past_opened_emails_count = last_user_emails.count { |msg| msg.opened_at.present? }
-    past_opened_emails_count / email_count
+    rand(min_day..max_day)
```
Comment (joshpuetz) [inline app/services/email_digest_article_collector.rb:52]: Since we can no longer use any calculations based on the open rate to determine when to send the next email digest, we'll just pick a random number between the min and max days specified in config.

**Comment by joshpuetz** [comment]
Putting this up for re-review as I've made some changes to the email digest sending code. Let me know what you think!

File: app/services/email_digest_article_collector.rb [hunk 5]
```diff
@@ -49,20 +49,7 @@ def days_until_next_email
     # Relies on hyperbolic tangent function to model the frequency of the digest email
```
Comment (rhymes) [inline app/services/email_digest_article_collector.rb:49]: this can be removed, we don't use `tanh` anymore

File: app/services/email_digest_article_collector.rb (hunk 4 above)
Comment (joshpuetz) [reply app/services/email_digest_article_collector.rb:51]: Great point! Any thoughts on this @forem/support ?

Comment (ellativity) [reply app/services/email_digest_article_collector.rb:51]: I'm not @michael-tharrington but it looks to me that relinquishing this data tracking (which I legit despise on a purely personal level) is worth the fix. But as I say, I'm biased and don't speak for the whole of Community Support.

And I'm also not Marketing so don't know how dependent we are on that data, besides timing email frequency.

Comment (michael-tharrington) [reply app/services/email_digest_article_collector.rb:51]: I'd agree with @ellativity that relinquishing the data tracking sounds worth the fix!

As far as Dev Digest emails go, everything I've heard is that folks receive them much more frequently than they'd expect. (see - https://github.com/forem/forem/issues/10940) So as long as this doesn't kick up the frequency of these emails then I think we're good.

**Comment by benhalpern** [comment]
The reason we track the open rate is to implement _exponential backoff_ for digest emails... I.e. if you open emails at a higher rate, you get something closer to the `min` rate, and if you don't open as often, we use something closer to the `max` rate.

While this was a pretty elegant solution, ultimately if we can't use it as it is today, we may want to do away with this complex logic altogether.

If we were to get rid of min and max, we'd reorient this to _only_ have a singular frequency rate functionality. That could radically simplify the logic we use to send these out. It would also mean we likely wouldn't even need `ahoy_email` _at all_.

`ahoy_email` is not responsible for any email functionality in any way, only for_tracking_. We only use the tracking for the above functionality.

The `ahoy_emails` table for DEV has... 35,041,804 rows... and it isn't exactly a table delivering a lot of value.

Email tracking was once fairly standard and unquestioned, but we're in a new era of privacy scrutiny, and we want to be on the right side of things. This is _a moment_, and this is why it is being removed from ahoy. We want to be on the right side of this story.

All that is to say, pending a conversation with product team, I think we could:

- Remove `ahoy_email` altogether
- Replace `...min` and `...max` with `periodic_email_digest_rate` which just defines that the digest goes out every x days.

This would be some extra work, but would save a huge amount of complexity for the system, admins, and end users. In the days of DEV as the main point of leverage, we needed to eek out every small bit of functionality which could help people come back and read posts on the site, but this type of DEV-specific precision for secondary is no longer as important relative to refinement of core features which deserve this kind of complexity.

Long story short, I think this may be an opportunity for ✂️ ✂️ ✂️... Let me circulate this and get clarity that we're cool with this, but I think it's likely the most pragmatic choice.

**Comment by Zhao-Andy** [comment]
Community Success does use the `EmailMessage` model -- which relies on the `ahoy_email` gem -- to see that we've sent emails as a response to abuse reports. I can discuss with the team later today and see if whether or not we want to keep it specifically for this feature.

**Comment by joshpuetz** [comment]
Putting this PR back into draft mode pending product decision mentioned in https://github.com/forem/forem/pull/13085#issuecomment-805271489

**Comment by joshpuetz** [comment]
Thanks @benhalpern, that's good context and I agree that removing `ahoy_email` would be a great step inline with our state goals and values!

Just and FYI, in addition to losing open email tracking, we'd also lose the following:
- recording message history of each email sent out
- automatic addition of UTM parameters on links inside digest emails, etc.

**Comment by Zhao-Andy** [comment]
Just a heads up from the Community Success meeting that we're happy to not be able to view emails this way anymore -- our quickest alternative is to make an automated note whenever an email is sent.

Also want to mention that we only have 4979 rows that are associated with a `FeedbackMessage`, so if we wanted to prune the table we would remove most of the emails that aren't used by Community Success.

**Comment by joshpuetz** [comment]
Soooooo....this is getting WAY more involved than the original 15 minute fix I planned for it to be! Here's what I propose:

- Finish up this PR by simplifying the digest frequency to a single value.
- Opening an issue to discuss removing this gem (and the additional work involved)

**Comment by benhalpern** [comment]
@joshpuetz that sounds fine to me. The ultimate solution will need an RFC, but if we're going to simplify down to just one frequency value we can start by doing that here.

We can also remove the _other_ types of tracking we currently do while you're at it:

These can all be false based on where we're headed with this.

```
AhoyEmail.default_options[:open] = true
AhoyEmail.default_options[:click] = true
AhoyEmail.default_options[:utm_params] = true
```

**Comment by djuber** [comment]
@joshpuetz it looks like the ahoy 2 gem was included but ahoy 1.1.1 was not removed (the delete was not in this changeset). 

I'll get a PR in a moment to include that.
